best_practices.rst

```diff
@@ -0,0 +1,474 @@
+services explicitly and simplifying the application maintenance.
+
+Use it in combination with :ref:`service autoconfiguration <services-autoconfigure>`
+to also add :doc:`service tags </service_container/tags>` to the services
+needing them, such as Twig extensions, event subscribers, etc.
+
+Services Should be Private Whenever Possible
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+:ref:`Make services private <container-public>` to prevent you from accessing
+those services via ``$container->get()``. Instead, you will need to use proper
+dependency injection.
+
+Use the YAML Format to Configure your Own Services
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If you use the :ref:`default services.yaml configuration <service-container-services-load-example>`,
+most services will be configured automatically. However, in some edge cases
+you'll need to configure services (or parts of them) manually.
+
+YAML is the format recommended to configure services because it's friendly to
+newcomers and concise, but Symfony also supports XML and PHP configuration, so
+you can use whatever format you like.
+
+Use Annotations to Define the Doctrine Entity Mapping
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Doctrine entities are plain PHP objects that you store in some "database".
+Doctrine only knows about your entities through the mapping metadata configured
+for your model classes.
+
+Doctrine supports several metadata formats, but it's recommended to use
+annotations because they are by far the most convenient and agile way of setting
+up and looking for mapping information.
+
+Controllers
+-----------
+
+Make your Controller Extend the ``AbstractController`` Base Controller
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Symfony provides a :ref:`base controller <the-base-controller-classes-services>`
+which includes shortcuts for the most common needs such as rendering templates
+or checking security permissions.
+
+Extending your controllers from this base controller couples your application
+to Symfony. Coupling is generally wrong, but it may be OK in this case because
+controllers shouldn't contain any business logic. Controllers should contain
+nothing more than a few lines of *glue-code*, so you are not coupling the
+important parts of your application.
+
+.. _best-practice-controller-annotations:
+
+Use Annotations to Configure Routing, Caching and Security
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Using annotations for routing, caching and security simplifies configuration.
+You don't need to browse tens of files created with different formats (YAML, XML,
+PHP): all the configuration is just where you need it and it only uses one format.
+
+The only exception is when the annotations are too complex, such as when using
+long expressions in the ``@Security`` annotation or lots of variables and
+requirements in the ``@Route`` annotation.
```

Comment:
I would remove this paragraph. For complex @Security expressions, use a voter instead. For complex routes, well, I don't think that it happens that often, so this is an edge case we don't need to talk about.